### ISSUE TITLE
[Paid] [$8000] zip /postcode validation error message not displayed for entering `,` on the Home address screen

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#14958]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. go to Settings > profile> personal details >Home address
2.  type the zip /postcode  as  ,

## Expected Result:
it should show validation error message

## Actual Result:
does not show validation message

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.2.67-7
**Reproducible in staging?:** y
**Reproducible in production?:** new feature
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/217662506-507818da-7e76-4a4a-b8b8-5525fc66e651.mov

https://user-images.githubusercontent.com/43996225/217662521-34c3d804-483d-4e42-a8fe-e87aaf5490fc.mp4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @gadhiyamanan 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1675838827973319

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~015b692899f4b1568b</li>
        <li>Upwork Job ID: 1623772333283323904</li>
        <li>Last Price Increase: 2023-03-02</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @anmurali (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by Beamanator** [comment]
A bit more context: on the "Personal Information" step of setting up a bank account show an error when entering `,` in the Zip code field. Noted in slack here: https://expensify.slack.com/archives/C049HHMV9SM/p1675844442499409?thread_ts=1675838827.973319&cid=C049HHMV9SM

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~015b692899f4b1568b

**Comment by MelvinBot** [comment]
Current assignee @anmurali is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @mananjadhav (`External`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @Beamanator (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by esh-g** [comment]
# Proposal

## Restating the problem
Some unwanted characters are allowed in zip code

## Root Cause
Regex is not being validated correctly for zip code

# Solution 

Here are some conditions for the post code:
- Every postal code system uses only A-Z and/or 0-9 and sometimes space/dash
- Not every country uses postal codes (ex. Ireland outside of Dublin), but we'll ignore that here.
- The shortest postal code format is Sierra Leone with NN
- The longest is American Samoa with NNNNN-NNNNNN
- You should allow one space or dash.
- Should not begin or end with space or dash

`^[a-z0-9][a-z0-9\- ]{0,10}[a-z0-9]$`
This regex should cover all those conditions
source: https://stackoverflow.com/a/19844362/15373641

As mentioned by @Beamanator in https://expensify.slack.com/archives/C049HHMV9SM/p1675845424815519?thread_ts=1675838827.973319&cid=C049HHMV9SM  We also have to allow letters like in Canadian zip codes, therefore allowing the alphabets

**Comment by abdulrahuman5196** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Validation errors are not shown in the address page with zip code

### What is the root cause of that problem?

We don't validate zip codes. In fact In the particular page i don't see any validation except empty check validation.
https://github.com/Expensify/App/blob/main/src/pages/settings/Profile/PersonalDetails/AddressPage.js#L116

### What changes do you think we should make in order to solve the problem?


We can add proper validation for all fields in the particular page including the concerned zip code.
We already have validations for these in other flows like company step. We can reuse the same functionality here as well.

For ex: We already have validation util for zipcode here - https://github.com/Expensify/App/blob/main/src/libs/ValidationUtils.js#L167

### What alternative solutions did you explore? (Optional)

**Reminder:** Please use plain English, be brief and avoid jargon. Feel free to use images, charts or pseudo-code if necessary. Do not post large multi-line diffs or write walls of text. Do not create PRs unless you have been hired for this job.

**Comment by topdev0119** [comment]
1. Develop a comprehensive plan for the migration that includes a timeline, budget, and resources needed.

2. Identify any potential risks associated with the migration and develop strategies to mitigate them.

3. Create a detailed inventory of all existing hardware, software, and data that needs to be migrated.

4. Develop a testing plan to ensure that all data is successfully migrated and that applications are functioning properly in the new environment.

5. Establish communication protocols between the old and new systems to ensure smooth transition of data and applications.

6. Train staff on how to use the new system and provide support during the transition period. 

7. Develop a backup plan in case of any unexpected issues during the migration process. 

8. Monitor progress throughout the migration process to ensure it is completed on time and within budget constraints.

**Comment by hackykitty** [comment]
## Proposal

### Re-state the problem  

zip/postcode validation error message not displayed for entering , on the Home address screen

### What is the root cause of that problem?

The root cause of this problem is the non-existent validation function for zip/postcode field.

### What changes do you think we should make in order to solve the problem?

Solution is very simple!!! Before clicking save button, just validate the zip/postcode field by using this function and show the validation error message if the user entered wrong. 
```
/**
 * @param {String} postcode 
 * @returns {Boolean}
 */
function isValidPostalCode(postcode) {
    return /^[0-9]{5}(?:-[0-9]{4})?$/.test(postcode);
}


``` 

### What alternative solutions did you explore? (Optional)

Optionally, we can use postal code API for postal code validation from https://developer.bring.com/api/postal-code/

**Comment by Changvang** [comment]
@twisterdotcom 
Hmm, Yeah, the vadationUtils.isRequiredFulfilled seem good with #15003 , but with this issue I don't think it not enough. Like it can't check zip code properly, like "    ,  " still pass check validate. You can check it.

 And with add some space to pass the validation of required field, I can see other some form get this issue like AddDebitCardPage.js, you see the billing address is required field but still not show required when I type more space. The validationUtils has using in this form, but not for address field, maybe other form like this scenario too.

<img width="417" alt="Screen Shot 2023-02-10 at 07 56 31" src="https://user-images.githubusercontent.com/33856037/217973087-704c1d08-a330-471c-bdfc-99782f04b6ea.png">

**Comment by mananjadhav** [comment]
@anmurali @Beamanator Can you please assign another C+ here? I have backlog of issues to cover from my OOO.

**Comment by situchan** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
`,` is allowed in zip / postcode input

### What is the root cause of that problem?
There's no validation check for zip / postcode field except `required` validation.
https://github.com/Expensify/App/blob/39d9b660e387b28688a221b976f0fbf52873cc67/src/pages/settings/Profile/PersonalDetails/AddressPage.js#L98-L123

### What changes do you think we should make in order to solve the problem?
I checked context:
> be aware that the current validation on that page only allows numbers (and an optional -) but in the new Home Address form we have to also allow letters since some countries have letters in their zip codes

So I think just allowing letters in current validation of zip code which was used in bank account step should be simple solution.
https://github.com/Expensify/App/blob/39d9b660e387b28688a221b976f0fbf52873cc67/src/CONST.js#L780
New zipcode regex for Home address would be:
`/[a-zA-Z0-9]{5}(?:[- ][a-zA-Z0-9]{4})?/`

Maybe we can define common function with `shouldAllowLetters` flag as a parameter and use this in both Home address page and Personal information page.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @aimane-chnaif (`External`)

**Comment by MelvinBot** [comment]
Current assignee @Beamanator is eligible for the External assigner, not assigning anyone new.

**Comment by anmurali** [comment]
@aimane-chnaif - I had to rotate contributor plus assignments since the original assignee didn't have bandwidth. Look forward to your review of the above proposals so we can pick one!

**Comment by MelvinBot** [comment]
@Beamanator, @anmurali, @aimane-chnaif Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by situchan** [comment]
## Proposal

[Updated](https://github.com/Expensify/App/issues/14958#issuecomment-1427806164) based on scope change https://github.com/Expensify/App/issues/14958#issuecomment-1431431974

**Comment by Beamanator** [comment]
Done 👍  @0xmiroslav can you please review proposals?

**Comment by MelvinBot** [comment]
@Beamanator @anmurali @0xmiroslav this issue was created 2 weeks ago. Are we close to approving a proposal? If not, what's blocking us from getting this issue assigned? Don't hesitate to create a thread in #expensify-open-source to align faster in real time. Thanks!

**Comment by Beamanator** [comment]
@0xmiroslav Looks like we came to a solution [in slack](https://expensify.slack.com/archives/C02NK2DQWUX/p1676995143307789?thread_ts=1676285113.004489&cid=C02NK2DQWUX)

**Comment by Beamanator** [comment]
> I think it would to be good to have this in common repo.

@0xmiroslav usually we only add things to the expensify-common repo if we also plan to immediately (or very soon) use the same logic in other repos / projects... At this exact moment we don't plan to add this validation to the OldDot / the current expenisfy.com website / app so I don't think we need it in expensify-common - can you think of another situation I'm missing?

**Comment by Beamanator** [comment]
> @Beamanator In that case, would we proceed with adding all this to the E/A repo? I could also potentially create a new library altogether and expose a function for this task, will reduce the unnecessary overhead of the lists.

Yes please let's keep it all in the E/App repo for now 👍  If you're thinking of creating a new library for this, do you mean a new file in the `src/libs` folder? We do have `ValidationUtils` which you may be able to add new functions to, maybe? Or if you are thinking of something more complex please explain 🙏 

> If we proceed with adding this in E/A, we should create a new file with all the data, I just want to avoid cluttering the `CONST.js` file.

Ooh since the file (https://github.com/FormatterKit/PostalCodeValidator/blob/master/Sources/PostalCodeValidator/PostalCodeValidator.swift) specifies regex by country iso, i wonder if we should hold on implementing country isos in this issue? https://github.com/Expensify/App/issues/14549

**Comment by MelvinBot** [comment]
@Beamanator @anmurali @0xmiroslav this issue is now 3 weeks old. There is one more week left before this issue <a href='https://stackoverflowteams.com/c/expensify/questions/15233/15234#15234'>breaks WAQ</a> and will need to go internal. What needs to happen to get a PR in review this week? Please create a thread in #expensify-open-source to discuss. Thanks!

**Comment by anmurali** [comment]
@Beamanator and @0xmiroslav are still evaluating proposals

**Comment by Beamanator** [comment]
@0xmiroslav I'm interested in hearing your feedback on the plans suggested so far, please let me know if you're unable to review the ideas today.

> > FYI I wouldn't worry too much about "cluttering the CONST.js file, it's already pretty ugly and I don't think a bunch more regexes would be thaaaat ugly
> 
> haha, sure. We could either:
> 
> 1. Proceed with adding the map to `CONST.js`.
> 2. Create a new file for storing the regex map.
> 3. Create a new npm package for this purpose as I mentioned above. No need to worry about maintenance since this essentially has no dependencies at all, just a simple magic-lamp package.

I think I'm ok with 1 or 2, but creating a new NPM package doesn't sound ideal

**Comment by Beamanator** [comment]
Hmm @0xmiroslav @Prince-Mendiratta I wonder if there's a website for generating values that satisfy certain regexes? If there is, that could probably be used to come up with examples / hints for each zip code, right? I don't think we need to make this too complicated.

Also @0xmiroslav I think we should go ahead and hire @Prince-Mendiratta for this job, then put this issue on HOLD while we wait for https://github.com/Expensify/App/issues/14549 to wrap up. Thoughts?

**Comment by MelvinBot** [comment]
📣 @Prince-Mendiratta You have been assigned to this job by @Beamanator!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by twisterdotcom** [comment]
The first part of the postal code for an address on `Tadcaster Road` would be `YO24` but a single address on the road would end with a secondary part, so for example `286 Tadcaster Rd York, York YO24 1ET` vs `131 Tadcaster Rd Dringhouses, Dringhouses, York YO24 1QJ`.

If they just entered `Tadcaster Road`, the Regex would correctly push them back, because the postal code isn't complete yet.

**Comment by MelvinBot** [comment]
@Beamanator, @anmurali, @Prince-Mendiratta, @0xmiroslav Huh... This is 4 days overdue. Who can take care of this?

**Comment by Beamanator** [comment]
Back-end change (https://github.com/Expensify/Web-Expensify/pull/36882) is on staging so PR can move forward!

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.2.94-3 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/15880


If no regressions arise, payment will be issued on 2023-04-12. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by MelvinBot** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@0xmiroslav / @Beamanator] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@0xmiroslav / @Beamanator] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@0xmiroslav / @Beamanator] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@anmurali] Determine if we should create a regression test for this bug.
- [ ] [@0xmiroslav] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@anmurali] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by MelvinBot** [comment]
@Beamanator, @anmurali, @Prince-Mendiratta, @0xmiroslav Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by Beamanator** [comment]
Bump @anmurali are you able to take care of this payment? I can assign another Bug-Zero team member if needed

**Comment by Beamanator** [comment]
Another quick bump @anmurali - I'm OOO till Thurs, I will ask for a new BZ if we haven't moved forward by then :D 👍

**Comment by anmurali** [comment]
Sent offers

**Comment by gadhiyamanan** [comment]
@anmurali accepted
Thanks!

**Comment by gadhiyamanan** [comment]
@anmurali please close the upwork contract
Thanks!

**Comment by MelvinBot** [comment]
@Beamanator, @anmurali, @Prince-Mendiratta, @0xmiroslav Huh... This is 4 days overdue. Who can take care of this?

**Comment by anmurali** [comment]
We're all done here!
